api/kv.go

```diff
@@ -0,0 +1,328 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+// A KVClient is used to perform reads and writes against a KV secrets engine in Vault.
+//
+// The mount path is the location where the target KV secrets engine resides
+// in Vault. The version refers to the version of the target KV secrets engine.
+//
+// Vault development servers tend to have "secret" as the mount path
+// and 2 as the version, as these are the default settings when a server
+// is started in -dev mode.
+//
+// Learn more about the KV secrets engine here:
+// https://www.vaultproject.io/docs/secrets/kv
+type KVClient struct {
+	c         *Client
+	mountPath string
+	version   int
+}
+
+type KVSecret struct {
+	Data     map[string]interface{}
+	Metadata *VersionMetadata
+	Raw      *Secret
+}
+
+type KVMetadata struct {
+	CASRequired        bool                   `mapstructure:"cas_required"`
+	CreatedTime        time.Time              `mapstructure:"created_time"`
+	CurrentVersion     int                    `mapstructure:"current_version"`
+	CustomMetadata     map[string]interface{} `mapstructure:"custom_metadata"`
+	DeleteVersionAfter time.Duration          `mapstructure:"delete_version_after"`
+	MaxVersions        int                    `mapstructure:"max_versions"`
+	OldestVersion      int                    `mapstructure:"oldest_version"`
+	UpdatedTime        time.Time              `mapstructure:"updated_time"`
+	// Keys are stringified ints, e.g. "3"
+	Versions map[string]VersionMetadata `mapstructure:"versions"`
```

Comment:
I couldn't find a way to make mapstructure happy if I had it as an int (since it comes in as a string), maybe there's some decode hook I could use but couldn't find it... 🤔  I did add a GetVersionsAsList method though so that folks have access to a sorted slice as well.